### PR TITLE
[docs] Managing ACL Policies

### PR DESCRIPTION
**Pull request by kaitlincarter-hc**
View the guide on Learn: https://deploy-preview-709--hashicorp-learn.netlify.com/consul/day-2-operations/managing-acl-policies

We created this guide from feedback from CS and HashiConf. The focus of the guide is to teach users how to discover the minimum privileges required to complete operations within your Consul datacenter and how to manage access using the operator-only implementation method.

### CODE REVIEW

File: website/source/docs/guides/managing-acl-policies.html.md [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+  }
+```
+
+## Secure Access Control: Operator-Only Access
+
+The most secure access control implementation restricts tokens with `acl=”write”` policies to only one or a few trusted operators. Tokens with the policy `acl = "write"` grant the holder unlimited privileges, because they can generate tokens with any other resource and policy. The operators are responsible for creating all other policies and tokens that grant limited access to the datacenter. We refer to this implementation as the operator-only implementation. This implementation type is the most secure, and most complex to manage.  
+
+For this implementation type operators are responsible for managing policies and tokens for: 
+- service registration
+- Connect proxy registration
```
Comment (banks) [inline website/source/docs/guides/managing-acl-policies.html.md]: I'm in two minds about whether it's more or less clear to call this out. Technically proxy registration is also just service registration but for a proxy type of service. But it may not be obvious to users that that is the case? The risk of calling them out separately is that it might imply that there is a different process in terms of ACL for these two when there kind of isn't.

I'm fine leaving it if you think it's clearer as I'm super on the fence!

File: website/source/docs/guides/managing-acl-policies.html.md [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+- intention management
+- agent management
+- API, CLI, and UI access
+
+For static or non-containerized workflows, this implementation type is straightforward and the operator's workload scales linearly. If the workflow is dynamic, implementing an automation tool will ensure the operator's workload does not scale exponentially.
```
Comment (banks) [inline website/source/docs/guides/managing-acl-policies.html.md]: 🤔 static vs dynamic doesn't really effect linear vs exponential scaling - in both cases the total work to be done in linear in the number of instances provisioned. It's more that in a dynamically scheduled environment, automation is _necessary_ to avoid having and manual operator step in deploying a new service. Even then it might be OK if you just distribute the same token to all instances (common although not ideal) - then the operator only is needed when a brand new service spins up.

I'd suggest maybe rewording this so we don't conflate scale of tokens needed with static vs dynamic requirement to provision new services/tokens?

File: website/source/docs/guides/managing-acl-policies.html.md [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+For this implementation type operators are responsible for managing policies and tokens for: 
+- service registration
+- Connect proxy registration
+- intention management
+- agent management
+- API, CLI, and UI access
+
+For static or non-containerized workflows, this implementation type is straightforward and the operator's workload scales linearly. If the workflow is dynamic, implementing an automation tool will ensure the operator's workload does not scale exponentially. 
+
+!> If you need to share token generation responsibilities with other teams, anyone with the ability to create tokens (`acl = "write"`) will also have unrestricted access to the datacenter. That individual will have the ability to deregister agents and services, and delete data from Consul KV.
```
Comment (banks) [inline website/source/docs/guides/managing-acl-policies.html.md]: I'd maybe add scarier warning here: acl:write basically makes you a super-user so you can bypass all security measures in ACL or Connect Intentions as well as delete any data.

```suggestion
!> If you need to share token generation responsibilities with other teams, anyone with the ability to create tokens (`acl = "write"`) effectively has unrestricted access to the datacenter. That individual will have the ability to access or delete any data in Consul including ACL tokens, Connect Intentions and all Catalog and KV data. 
```

File: website/source/docs/guides/managing-acl-policies.html.md [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+
+For static or non-containerized workflows, this implementation type is straightforward and the operator's workload scales linearly. If the workflow is dynamic, implementing an automation tool will ensure the operator's workload does not scale exponentially. 
+
+!> If you need to share token generation responsibilities with other teams, anyone with the ability to create tokens (`acl = "write"`) will also have unrestricted access to the datacenter. That individual will have the ability to deregister agents and services, and delete data from Consul KV. 
+
+### Operator-Only Implementation Example 
+
+In this following example Operators retain responsibility for service token management, but delegate access control between Connect-enabled services to the security team. 
+
+-> Note: The service registration examples describe the token generation process for per-service tokens, therefore, they are only applicable if you are able to create per-service tokens.
```
Comment (banks) [inline website/source/docs/guides/managing-acl-policies.html.md]: I don't really follow this note - isn't it a tautology that the instructions only apply if they described the situation I'm in? Maybe I'm missing something?

File: website/source/docs/guides/managing-acl-policies.html.md [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+- agent management
+- API, CLI, and UI access
+
+For static or non-containerized workflows, this implementation type is straightforward and the operator's workload scales linearly. If the workflow is dynamic, implementing an automation tool will ensure the operator's workload does not scale exponentially. 
+
+!> If you need to share token generation responsibilities with other teams, anyone with the ability to create tokens (`acl = "write"`) will also have unrestricted access to the datacenter. That individual will have the ability to deregister agents and services, and delete data from Consul KV. 
+
+### Operator-Only Implementation Example 
+
+In this following example Operators retain responsibility for service token management, but delegate access control between Connect-enabled services to the security team.
```
Comment (banks) [inline website/source/docs/guides/managing-acl-policies.html.md]: "security team" not capitalized here but is elsewhere.

File: website/source/docs/guides/managing-acl-policies.html.md [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+
+!> If you need to share token generation responsibilities with other teams, anyone with the ability to create tokens (`acl = "write"`) will also have unrestricted access to the datacenter. That individual will have the ability to deregister agents and services, and delete data from Consul KV. 
+
+### Operator-Only Implementation Example 
+
+In this following example Operators retain responsibility for service token management, but delegate access control between Connect-enabled services to the security team. 
+
+-> Note: The service registration examples describe the token generation process for per-service tokens, therefore, they are only applicable if you are able to create per-service tokens. 
+
+Initially, the Operator creates a single token with intention management privileges for the Security Team, and another service token for the Developer. For intention management the `intention` policy disposition should be included in the service rule.
```
Comment (banks) [inline website/source/docs/guides/managing-acl-policies.html.md]: nit: You spoke about two tokens one for sec team one for dev but then didn't explicitly specify which one you mean in the second sentence. It's perfectly clear if you follow it since you said "intention management" was for "security" in the last sentence but I wonder if it could be even clearer if you removed the mention of a second token or made the second sentence more explicit?

For example:

> Initially, the Operator creates a single token with intention management privileges for the Security Team. Developers will be given separate tokens to enable them to register their applications without having access to manage Intentions.
>
> The security team's token must include the `intention` policy disposition in the service rule for any services they should be allowed to manage Intentions for.

File: website/source/docs/guides/managing-acl-policies.html.md [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+
+```hcl
+service "wordpress" 
+  { policy = "read" 
+    intentions = "write"
+  }
+```
+
+This enables the security team to create an intention that allows the upstream proxy for `mysql` to send data to the destination, the downstream proxy for `wordpress`.
```
Comment (banks) [inline website/source/docs/guides/managing-acl-policies.html.md]: Suggest removing proxies here: they are an implementation detail and this rule works just the same if apps are not using proxies at all. I think people will understand that proxies are just the transport and that the intention is between the two services.

Also this example is backwards - wordpress is the client and needs to establish connections to an _upstream_ Mysql DB so it should be `mysql` in the rule above and this sentence reversed:

```suggestion
This enables the security team to create an intention that allows the `wordpress` service to open new connections to the upstream `mysql` service.
```

File: website/source/docs/guides/managing-acl-policies.html.md [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+
+```sh
+consul intention create wordpress mysql
+``` 
+
+The Security Team responsible for managing intentions would need to create allow intentions when the `default_policy` is set to `deny`, since deny all will be inherited from the ACL configuration.
+
+In this implementation type, Developers are responsible for requesting a token for their service. For a Connect enabled service, the Operator would need to create a policy that provides write privileges for the service and proxy, and read privileges for the service_prefix and node_prefix.
```
Comment (banks) [inline website/source/docs/guides/managing-acl-policies.html.md]:
```suggestion
In this implementation type, Developers are responsible for requesting a token for their service. For a Connect enabled service, the Operator would need to create a policy that provides write privileges for the service and proxy, and read privileges for all services and nodes to enable discovery of other upstream dependencies.
```

Maybe worth adding a note here that it's possible to be more restrictive with service and node read privs but it quickly adds complexity to whitelist all the services and nodes they run on that a token should be allowed to discover upstreams via DNS, API or Connect.

File: website/source/docs/guides/managing-acl-policies.html.md [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+```hcl
+service "mysql" {
+  policy = "write"
+}
+
+service "mysql-sidecar-proxy" {
+  policy = "write"
+}
+
+service_prefix "" {
```
Comment (banks) [inline website/source/docs/guides/managing-acl-policies.html.md]: Is it worth being explicit about the fact that an empty prefix matches everything?

File: website/source/docs/guides/managing-acl-policies.html.md [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+
+```sh
+$ consul services register mysql
+``` 
+
+The Operator example above illustrates creating policies on the security spectrum. The first example, using an exact match resource rule, is the most secure. It grants the least privileges necessary to accomplish the task. Generally, creating policies and tokens with the least privileges will result in more policy definitions. However, this will help you create the most secure environment. Alternatively, the prefix rule can apply to zero-to-many objects. The trade-off of a less complicated token creation process is security. Note, this applies to all rules not just for agents.  
+
+## Next Steps
+
+After setting up access control processes, you will need to implement a token rotation policy. If you are using third-party tool to generate tokens, such as Vault, Consul ACL tokens will adhere to the TTLs set in that third party tool. If you are manually rotating tokens or need to revoke access, you can delete a token at any time with the [API](https://www.consul.io/api/acl/tokens.html#delete-a-token).
```
Comment (banks) [inline website/source/docs/guides/managing-acl-policies.html.md]: Worth noting here that if you do use automation to rotate tokens, then the application or tooling also needs to re-register with the local agent whenever the token changes.

Currently we don't document that anywhere so people who use the Vault integration will have their tokens expire after 30 days but typically never re-register with a new one meaning the consul agent actually looses the ability to register updates or update health status for the instance! In Connect world this is even worse as your certificates will not be able to be renewed after this and all traffic will break etc!

We should at least mention here when talking about rotation that currently rotation requires re-regristration with the new token to provide the local agent with a refreshed token to continue to work with.

File: website/source/docs/guides/managing-acl-policies.html.md [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+content_length: 15
+id: managing-acl-policies
+products_used:
+  - Consul
+description: >-
+  In this guide, you'll learn how to discover the minimum privileges required to complete operations within your Consul datacenter and how to manage access using the operator-only implementation method.
+level: Implementation
+---
+
+This guide is for Operators with the responsibility of creating and managing ACL tokens for a Consul datacenter. It includes several recommendations on how to discover the minimum privileges required to complete operations. Throughout the guide we'll provide examples and use cases that you can adapt to your environment, however, it does not include environment specific recommendations. After completing this guide, you will have a better understanding of how to effectively manage ACL policies and tokens.
```
[the remark this answers is not in the document]
Comment (kaitlincarter-hc) [reply website/source/docs/guides/managing-acl-policies.html.md]: Saying "Consul datacenter" seems unnecessary actually.

File: website/source/docs/guides/managing-acl-policies.html.md [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+
+## Discover Required Privileges 
+
+After bootstrapping the ACL system and configuring Consul agents with tokens, you will need to create tokens to complete any additional task within the datacenter including registering services. 
+
+Before discovering the minimum privileges, it's important to understand the basic components of a token. A rule is a specific privilege and the basic unit of a token. Rules are combined to create policies. There are two main parts of a rule, the resource and policy disposition. The resource is the object that the rule applies to and the policy dispositions dictates privileges. The example below applies to any service object named "web". The policy disposition grants read privileges. 
+
+![ACL Rule Diagram](/static/img/consul/ACL-rule.png 'ACL Diagram with rules')
```
[the remark this answers is not in the document]
Comment (kaitlincarter-hc) [reply website/source/docs/guides/managing-acl-policies.html.md]: I've updated it to a call out box!

File: website/source/docs/guides/managing-acl-policies.html.md (hunk 9 above)
Comment (kaitlincarter-hc) [reply website/source/docs/guides/managing-acl-policies.html.md]: Done!

File: website/source/docs/guides/managing-acl-policies.html.md [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+
+To discover the minimum privileges required for a specific operation, we have three recommendations. 
+
+First, focus on the data in your environment that needs to be secured. Ensure your sensitive data has policies that are specific and limited. Since policies can be combined to create tokens, you will usually write more policies for sensitive data. Sensitive data could be a specific application or a set of values in the key-value store. 
+
+Second, reference the Consul docs, both the [rules page](https://www.consul.io/docs/acl/acl-rules.html) and [API pages](https://www.consul.io/api/index.html), often to understand the required privileges for any given operation. 
+
+The rules documentation explains the 11 rule resources. The following four resource types are critical for any operating datacenter with ACLs enabled.
```
[the remark this answers is not in the document]
Comment (kaitlincarter-hc) [reply website/source/docs/guides/managing-acl-policies.html.md]: Thats a great point. I'll update it!

File: website/source/docs/guides/managing-acl-policies.html.md (hunk 6 above)
Comment (kaitlincarter-hc) [reply website/source/docs/guides/managing-acl-policies.html.md]: Updated!

File: website/source/docs/guides/managing-acl-policies.html.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (kaitlincarter-hc) [reply website/source/docs/guides/managing-acl-policies.html.md]: I removed "Consul datacenter" since it felt a little redundant, since you're correct, managing ACLs will apply to one or many datacenters.

File: website/source/docs/guides/managing-acl-policies.html.md [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+|`node` and  `node_prefix` |Node rules grant privileges for node-level registration, including adding agents to the datacenter and catalog.
+| `service` and `service_prefix`   | Service rules grant privileges for service-level registration, including adding services to the catalog. 
+| `operator`        | The operator grants privileges for datacenter operations, including interacting with Raft. 
+
+
+On the API pages, each endpoint will have a table that includes required ACLs. The node health endpoint, shown below, requires node and service read to view all checks for the specified node. 
+
+![API Health Endpoint](/static/img/consul/api-endpoint.png 'Screenshot of Health endpoint page')
+
+Finally, before using a token in production, you should test that it has the correct privileges. If the token being used is missing a privilege, then Consul will issue a 403 permission denied error.
```
[the remark this answers is not in the document]
Comment (kaitlincarter-hc) [reply website/source/docs/guides/managing-acl-policies.html.md]: Noted!

File: website/source/docs/guides/managing-acl-policies.html.md [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+Finally, before using a token in production, you should test that it has the correct privileges. If the token being used is missing a privilege, then Consul will issue a 403 permission denied error. 
+
+```sh
+$ consul operator raft list-peers
+Error getting peers: Failed to retrieve raft configuration: Unexpected response code: 403 (rpc error making call: Permission denied)
+``` 
+
+### Operations Example: Listing Consul Agents
+
+To view all Consul agents in the datacenter, you can use the `consul members` CLI command. You will need at least read privileges for every agent in the datacenter. Depending on your threat model you could have either of the following policies.
```
[the remark this answers is not in the document]
Comment (kaitlincarter-hc) [reply website/source/docs/guides/managing-acl-policies.html.md]: node vs agent privileges are really confusing and I refer back to our rule docs often!

File: website/source/docs/guides/managing-acl-policies.html.md [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+--------------------|--------------------------------------------------------
+|`acl`              |ACL rules grant privileges for ACL operations including to create,update, or view tokens and policies.
+|`node` and  `node_prefix` |Node rules grant privileges for node-level registration, including adding agents to the datacenter and catalog.
+| `service` and `service_prefix`   | Service rules grant privileges for service-level registration, including adding services to the catalog. 
+| `operator`        | The operator grants privileges for datacenter operations, including interacting with Raft. 
+
+
+On the API pages, each endpoint will have a table that includes required ACLs. The node health endpoint, shown below, requires node and service read to view all checks for the specified node. 
+
+![API Health Endpoint](/static/img/consul/api-endpoint.png 'Screenshot of Health endpoint page')
```
[the remark this answers is not in the document]
Comment (kaitlincarter-hc) [reply website/source/docs/guides/managing-acl-policies.html.md]: Done!

File: website/source/docs/guides/managing-acl-policies.html.md [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+```sh
+$ consul operator raft list-peers
+Error getting peers: Failed to retrieve raft configuration: Unexpected response code: 403 (rpc error making call: Permission denied)
+``` 
+
+### Operations Example: Listing Consul Agents
+
+To view all Consul agents in the datacenter, you can use the `consul members` CLI command. You will need at least read privileges for every agent in the datacenter. Depending on your threat model you could have either of the following policies. 
+
+An individual rule and policy for each agent: this requires you to add a new policy to the token every time a new agent is added to the datacenter. Your token would have as many policies as agents in the datacenter. This method is ideal for a static environment, and for maximum security.
```
[the remark this answers is not in the document]
Comment (kaitlincarter-hc) [reply website/source/docs/guides/managing-acl-policies.html.md]: This example is trying to highlight the usability vs security trade-offs. Also, I tried to make it clear in the intro that we wouldn't expect operators to make a manual update, but rather they'd automate the processes. 
I updated this section to be a little less prescriptive!

File: website/source/docs/guides/managing-acl-policies.html.md (hunk 1 above)
Comment (kaitlincarter-hc) [reply website/source/docs/guides/managing-acl-policies.html.md]: I'm putting them on the same line! My original thought was sometimes users will register the proxy and service separately. Also, even though the registration process is the same users still need to actively think about registering proxies since its not automatic.
